### PR TITLE
quick fix for specnlm2spec1p

### PR DESCRIPTION
**Pull request by zhanglw0521**
cf issue #28

### CODE REVIEW

**Comment by cortner** [comment]
Do you understand why we implemented it with a `union` in the past? If not, then please leave the original implementation, comment it out (just the one line) and add a remark that we don't understand why the original implementation was done this way. Then we can more easily recover it if a different use-case fails in the near future.

**Comment by cortner** [comment]
one possibility is that it's just a matter of size of the input array - would you mind quickly testing this?

**Comment by zhanglw0521** [comment]
> Do you understand why we implemented it with a `union` in the past? If not, then please leave the original implementation, comment it out (just the one line) and add a remark that we don't understand why the original implementation was done this way. Then we can more easily recover it if a different use-case fails in the near future.

Sure, will do. I remember it was me who used the `union` function, just because it did the job and I am guessing that it was the ```...``` thing that mattered. The new code in this PR does exactly the same thing as the old one but I will just comment out the old version instead of deleting it to make things recoverable. 

> one possibility is that it's just a matter of size of the input array - would you mind quickly testing this?

And yes, it is indeed related to the size of the input array. In the following 
```julia
A = [ [rand(1:300) for _ = 1:3] for __ = 1:Nmax]
union(A...)
```
the bound for `Nmax` for this error to show up is roughly 20000.

**Comment by cortner** [comment]
ok if it's size of the input then it's clear what is happening. The `...` operator creates a new argument for every element of the array. This quickly goes outside stack size hence the overflow. Can't you just use the `union` with the splatting?

**Comment by zhanglw0521** [comment]
Perhaps we can not (just using `union` without splatting), or at least not so direct I guess. `AAspec` is a vector of vectors of `Aspec`. if we use `union(AAspec)`, it is more or less the same as `unique(AAspec)` (and in our case, this is still AAspec itself as it contains distinct elements already), which is still a vector of vectors of `Aspec` but not a set of all used `Aspec`s.

File: src/utils.jl [hunk 1]
```diff
@@ -154,7 +154,12 @@ specnlm2spec1p(spec_nlm)
 From a list of AA specifications to all A specifications needed
 """
 function specnlm2spec1p(spec_nlm)
-    spec1p = union(spec_nlm...)
+    # spec1p = union(spec_nlm...) # oringinal implementation which leads to StackOverflowError when input vector is too long
+    spec1p = []
```
Comment (cortner) [inline src/utils.jl:158]: could you initilize this with `spec1p = spec_nlm[1]`, then you get type stability

File: src/utils.jl [hunk 2]
```diff
@@ -154,7 +154,12 @@ specnlm2spec1p(spec_nlm)
...
+    for spec_nlm_i in spec_nlm
+        push!(spec1p, spec_nlm_i...)
+        unique!(spec1p)
+    end
     lmax = [ spec1p[i].l for i = 1:length(spec1p) ] |> maximum
     nmax = [ spec1p[i].n for i = 1:length(spec1p) ] |> maximum
     return spec1p, lmax, nmax + 1
```
Comment (cortner) [inline src/utils.jl:165]: an alternative to get type stability is to return `identity.(spec1p)`

**Comment by cortner** [comment]
thanks - a few more minor comment and then feel free to merge yourself.

**Comment by zhanglw0521** [comment]
Thanks very much for the detailed review. I will modify and merge in the next few minutes.

File: src/utils.jl (hunk 2 above)
Comment (zhanglw0521) [reply src/utils.jl:165]: I'll follow this one

**Pull request merged**